Rectangle.js

```diff
@@ -0,0 +1,157 @@
+        let secondCenterY = second.y + second.height / 2;
+        let collideX = this.findDistance(first.width, second.width, firstCenterX, secondCenterX, 0, false);
+        let collideY = this.findDistance(first.height, second.height, firstCenterY, secondCenterY, 0, false);
+        return collideX && collideY;
+    }
+
+    checkSticky(first, second) {
+        let firstCenterX = first.x + first.width / 2;
+        let firstCenterY = first.y + first.height / 2;
+        let secondCenterX = second.x + second.width / 2;
+        let secondCenterY = second.y + second.height / 2;
+        let stickyX = this.findDistance(first.width, second.width, firstCenterX, secondCenterX, N, false);
+        let stickyY = this.findDistance(first.height, second.height, firstCenterY, secondCenterY, N, false);
+        let collideX = this.findDistance(first.width, second.width, firstCenterX, secondCenterX, 0, true);
+        let collideY = this.findDistance(first.height, second.height, firstCenterY, secondCenterY, 0, true);
+        if ((firstCenterY < secondCenterY) && stickyX && stickyY && collideY) {
+            return 1;
+        }
+        if ((firstCenterX > secondCenterX) && stickyX && stickyY && collideX) {
+            return 2;
+        }
+        if ((firstCenterY > secondCenterY) && stickyX && stickyY && collideY) {
+            return 3;
+        }
+        if ((firstCenterX < secondCenterX) && stickyX && stickyY && collideX) {
+            return 4;
+        }
+        return false;
+    }
+
+    isSticky(rect) {
+        if (!this.stickedTo) {
+            switch (this.checkSticky(this, rect)) {
+                case 1:
+                    this.stickySide = 1;
+                    this.stickedTo = rect;
+                    break;
+                case 2:
+                    this.stickySide = 2;
+                    this.stickedTo = rect;
+                    break;
+                case 3:
+                    this.stickySide = 3;
+                    this.stickedTo = rect;
+                    break;
+                case 4:
+                    this.stickySide = 4;
+                    this.stickedTo = rect;
+                    break;
+                default:
+                    break;
+            }
+        } else {
+            if (this.stickedTo === rect) {
+                if (!this.checkSticky(this, rect)) {
+                    this.stickedTo = null;
+                    this.stickySide = null;
+
+                }
+            }
+        }
+    }
+
+    isCollided(rect) {
+        if (rect !== this) {
+            if (rect) {
+                return this.checkCollision(this, rect);
+            } else {
+                return this.collided;
+            }
+        }
+    }
+
+    draw(context) {
+        context.beginPath();
+        if (this.stickedTo) {
+            switch (this.stickySide) {
+                case 1:
+                    this.y = this.stickedTo.y - this.height;
```

Comment:
`draw` method should only draw existing state on canvas, not introduce any side-effects by modifying something